common/workunit/workflow.cpp

```diff
@@ -613,12 +804,867 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+    for(int i = 1; i <= workflow->count(); i++)
+    {
+        CCloneWorkflowItem & cur = queryWorkflowItem(i);
+
+        if((cur.queryNumPredecessors() == 0) && cur.isActive())
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "item %u has been added to the inital task queue", cur.queryWfid());
+        #endif
+            wfItemQueue.push(i);
+        }
+    }
+    wfItemQueueSem.signal(wfItemQueue.size());
+}
+bool WorkflowMachine::checkIfDone()
+{
+    if((activeContingencies == 0) && (parentReached))
+    {
+#ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "WorkflowMachine::checkifDone. Final check. Branch count: %u", branchCount.load());
+#endif
+        if((branchCount == 0))
+        {
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "workflow done");
+    #endif
+            done = true;
+            wfItemQueueSem.signal(numThreads);
+            return true;
+        }
+    }
+    return false;
+}
+void WorkflowMachine::processWfItems()
+{
+    while(!done)
+    {
+        wfItemQueueSem.wait();
+        if(!done)
+        {
+            unsigned currentWfid = 0;
+            //enter critical section
+            {
+                CriticalBlock thisBlock(queueCritSec);
+                currentWfid = wfItemQueue.front();
+                wfItemQueue.pop();
+            }
+            try
+            {
+                executeItemParallel(currentWfid);
+            }
+            //terminate threads on fatal exception and save error
+            catch(WorkflowException * e)
+            {
+                runtimeError.setown(e);
+                done = true;
+                wfItemQueueSem.signal(numThreads); //MORE: think about interrupting other threads
+                break;
+            }
+        }
+    }
+}
+void WorkflowMachine::performParallel(IGlobalCodeContext *_ctx, IEclProcess *_process)
+{
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "starting perform parallel");
+#endif
+    ctx = _ctx;
+    process = _process;
+
+    //relink workflow
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Starting to mark Items with their successors");
+#endif
+    addSuccessors();
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Finished marking Items with their successors");
+#endif
+
+    //add initial values
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Adding initial workflow items");
+#endif
+    initialiseItemQueue();
+
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Initialising threads");
+#endif
+
+    //initialise thread count
+    numThreads = getThreadNumFlag();
+    if(numThreads < 1)
+        numThreads = 4;
+    unsigned maxThreads = getAffinityCpus();
+    if(numThreads > maxThreads)
+        numThreads = maxThreads;
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "num threads = %u", numThreads);
+#endif
+
+    std::vector<std::thread *> threads(numThreads);
+    //Start threads
+    for(int i=0; i < numThreads; i++)
+        threads[i] = new std::thread([this]() {  this->processWfItems(); });
+
```

Comment:
I would add a comment here:  
//NOTE: Initial work items have already been added to the queue by initializeItemQueue() above.